### ISSUE TITLE
Global StringCache does not work when creating dataframes with categorical data directly.

### ISSUE THREAD

**Issue by ghuls** [pola-rs/polars#627]
#### Are you using Python or Rust?

**Python**

#### What version of polars are you using?
0.7.15

#### What operating system are you using polars on?

Ubuntu 20.04

#### Describe your bug.

Global StringCache does not work when creating dataframes with categorical data directly as opposed to first creating the dataframe and changing a column to Categorical.

#### What are the steps to reproduce the behavior?

This works:

```python
with pl.StringCache():
    df1 = pl.DataFrame({"region_ids": ["reg1", "reg2", "reg3", "reg4", "reg5"]})
    df2 = pl.DataFrame({"seq_name": ["reg4", "reg2", "reg1"], "score": [3.0, 1.0, 2.0]})
    df1["region_ids_cat"] = df1["region_ids"].cast(pl.Categorical)
    df2["seq_name"] = df2["seq_name"].cast(pl.Categorical)

print(df1.join(df2, left_on='region_ids_cat', right_on='seq_name', how='left'))
```

Output:

```
shape: (5, 3)
╭────────────┬────────────────┬───────╮
│ region_ids ┆ region_ids_cat ┆ score │
│ ---        ┆ ---            ┆ ---   │
│ str        ┆ cat            ┆ f64   │
╞════════════╪════════════════╪═══════╡
│ "reg1"     ┆ "reg1"         ┆ 2     │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
│ "reg2"     ┆ "reg2"         ┆ 1     │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
│ "reg3"     ┆ "reg3"         ┆ null  │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
│ "reg4"     ┆ "reg4"         ┆ 3     │
├╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌┤
│ "reg5"     ┆ "reg5"         ┆ null  │
╰────────────┴────────────────┴───────╯
```

Creating a dataframe and setting a column to Categorical directly, does not use the StringCache.

```python
with pl.StringCache():
    df1 = pl.DataFrame({"region_ids": ["reg1", "reg2", "reg3", "reg4", "reg5"]}).select([pl.col("region_ids").cast(pl.Categorical)])
    df2 = pl.DataFrame({"seq_name": ["reg4", "reg2", "reg1"], "score": [3.0, 1.0, 2.0]}).select([pl.col("seq_name").cast(pl.Categorical), pl.col("score")])

print(df1.join(df2, left_on='region_ids', right_on='seq_name', how='left'))
```

Output:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-3-0a339f0b20d2> in <module>
----> 1 print(df1.join(df2, left_on='region_ids', right_on='seq_name', how='left'))

~/software/anaconda3/envs/create_cistarget_databases/lib/python3.8/site-packages/polars/frame.py in join(self, df, left_on, right_on, on, how)
   1222             return self.lazy().join(df.lazy(), left_on, right_on, how=how)
   1223 
-> 1224         out = self._df.join(df._df, left_on, right_on, how)
   1225 
   1226         return wrap_df(out)

RuntimeError: Any(ValueError("joins on categorical dtypes can only happen if they are created under the same global string cache"))
```

Also assigning new values to a dataframe with one column does not seem to work (Creating a new column works with this syntax: `df1["region_ids_cat"] = df1["region_ids"].cast(pl.Categorical)`, but `df1["region_ids"] = df1["region_ids"].cast(pl.Categorical)` does not.

```python
with pl.StringCache():
    df1 = pl.DataFrame({"region_ids": ["reg1", "reg2", "reg3", "reg4", "reg5"]})
    df2 = pl.DataFrame({"seq_name": ["reg4", "reg2", "reg1"], "score": [3.0, 1.0, 2.0]})
    df1["region_ids"] = df1["region_ids"].cast(pl.Categorical)
    df2["seq_name"] = df2["seq_name"].cast(pl.Categorical)

print(df1.join(df2, left_on='region_ids_cat', right_on='seq_name', how='left'))
```

Output:

```
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
<ipython-input-6-8f7c4055980a> in <module>
      2     df1 = pl.DataFrame({"region_ids": ["reg1", "reg2", "reg3", "reg4", "reg5"]})
      3     df2 = pl.DataFrame({"seq_name": ["reg4", "reg2", "reg1"], "score": [3.0, 1.0, 2.0]})
----> 4     df1["region_ids"] = df1["region_ids"].cast(pl.Categorical)
      5     df2["seq_name"] = df2["seq_name"].cast(pl.Categorical)
      6 

~/software/anaconda3/envs/create_cistarget_databases/lib/python3.8/site-packages/polars/frame.py in __setitem__(self, key, value)
    587             except Exception:
    588                 pass
--> 589             self.hstack([Series(key, value)], in_place=True)
    590         # df[idx] = series
    591         elif isinstance(key, int):

~/software/anaconda3/envs/create_cistarget_databases/lib/python3.8/site-packages/polars/frame.py in hstack(self, columns, in_place)
   1273             columns = columns.get_columns()
   1274         if in_place:
-> 1275             self._df.hstack_mut([s.inner() for s in columns])
   1276         else:
   1277             return wrap_df(self._df.hstack([s.inner() for s in columns]))

RuntimeError: Any(ShapeMisMatch("Could not horizontally stack Series. The Series length 5 differs from the DataFrame height: 0"))
```

**Comment by ritchie46** [comment]
Thank you for the issue report, I will look into it.

**Comment by ritchie46** [comment]
This one is fixed in #628
